lci_strategic_plugin/src/lci_strategic_plugin.cpp

```diff
@@ -608,37 +608,66 @@ void LCIStrategicPlugin::planWhenAPPROACHING(const cav_srvs::PlanManeuversReques
     
     // check if scheduled_enter_time_ is inside the available states interval
     size_t i = 0;
+    
+
     for (auto pair : traffic_light->recorded_time_stamps)
     {
-      if (pair.second == lanelet::CarmaTrafficSignalState::PROTECTED_MOVEMENT_ALLOWED && 
-        lanelet::time::timeFromSec(nearest_green_entry_time.toSec()) < pair.first ) 
+      if (lanelet::time::timeFromSec(nearest_green_entry_time.toSec()) < pair.first)
       {
-        ROS_DEBUG_STREAM("ET is inside the available future events! where starting time: " << std::to_string(lanelet::time::toSec(traffic_light->recorded_start_time_stamps[i])) 
-          << ", ending time of that green signal is: " << std::to_string(lanelet::time::toSec(pair.first)));
-        is_entry_time_within_future_events = true;
+        if (pair.second == lanelet::CarmaTrafficSignalState::PROTECTED_MOVEMENT_ALLOWED) 
+        {
+          ROS_DEBUG_STREAM("ET is inside the GREEN phase! where starting time: " << std::to_string(lanelet::time::toSec(traffic_light->recorded_start_time_stamps[i])) 
+            << ", ending time of that green signal is: " << std::to_string(lanelet::time::toSec(pair.first)));
+          is_entry_time_within_green_or_tdb = true;
+        }
+        else
+        {
+          ROS_ERROR_STREAM("Vehicle should plan cruise and stop as ET is inside the RED or YELLOW phase! where starting time: " << std::to_string(lanelet::time::toSec(traffic_light->recorded_start_time_stamps[i])) 
+            << ", ending time of that green signal is: " << std::to_string(lanelet::time::toSec(pair.first)));
+          is_entry_time_within_green_or_tdb = false;
+        }
+
+        in_tdb = false;
         break;
       }
       i++;
     }
+
+    if (in_tdb)
+    {
+      ROS_DEBUG_STREAM("ET is inside TDB phase! where starting time: " << std::to_string(lanelet::time::toSec(traffic_light->recorded_time_stamps.back().first)));
+      is_entry_time_within_green_or_tdb = true;
+    }
+
   }
+
+ 
   
   ROS_DEBUG_STREAM("nearest_green_entry_time: " << std::to_string(nearest_green_entry_time.toSec()) << ", with : " << nearest_green_entry_time - current_state.stamp  << " seconds left at: " << std::to_string(current_state.stamp.toSec()));
   
-  if (!nearest_green_entry_time_cached_ && is_entry_time_within_future_events) 
+  if (nearest_green_entry_time_cached_) 
+  { // always pick later of buffered green entry time, or earliest entry time
+    nearest_green_entry_time = ros::Time(std::max(nearest_green_entry_time.toSec(), nearest_green_entry_time_cached_.get().toSec()));
+  }
+
+  ROS_DEBUG_STREAM("After accounting for cached - nearest_green_entry_time: " << std::to_string(nearest_green_entry_time.toSec()) << ", with : " << nearest_green_entry_time - current_state.stamp  << " seconds left at: " << std::to_string(current_state.stamp.toSec()));
+  
+  if (!nearest_green_entry_time_cached_ && is_entry_time_within_green_or_tdb) 
```

Comment:
I don't understand why we need the first one? I am guessing what you want to say is if nearest_green_entry_time = nearest_green_entry_time_cached_, then none of these are required to check. Note that, it is possible that nearest_green_entry_time is not overwritten by nearest_green_entry_time_cached_. Anyway, I would say let remove the first check. If nearest_green_entry_time_cached_ exist, it will be checked earlier when calculating nearest_green_entry_time. So I don't see it necessary here. 

Also, you want to have an else condition for this check. If is_entry_time_within_green_or_tdb not true, then nearest_green_entry_time is within red or yellow, which means the vehicle shall go to handle stopping case. How are we capturing this case?

Reply:
- Essentially on a higher level, this check is saying if the "logic to apply green_buffer" has been done before, avoid doing it again. Basically having nearest_green_entry_time_cached_ means buffer applying logic is run before and is good to reuse again. And this cached is only specific to one ET from street and resets if given a new one as mentioned in the previous comment.
- if not in green_or_tdb (which LCI now logs before this logic with ERROR msgs as requested), no buffer will be applied and the raw ET will be used to calculate some TS, and that TS will be failed by canArriveWithCertainty check which will make it cruiseAndStop.

Reply:
- I still thing the first check is not necessary. It is possible that nearest_green_entry_time_cached_ is not empty, but nearest_green_entry_time is greater than nearest_green_entry_time_cached_, and therefore, whether we need to add buffer or not needs to be checked again. With the implemented logic, nothing will be checked in this case, which is wrong.
- I see now. Ok sounds good.

Reply:
nearest_green_entry_time_cached_ is minimum nearest_green_start_time + buffer. So if nearest_green_entry_time > nearest_green_entry_time_cached_ , automatically we don't need to apply buffer.

Reply:
Sort of true. I guess the only case nearest_green_start_time is greater than nearest_green_entry_time_cached_ is when EET is greater than nearest_green_entry_time_cached_. In this case, I guess EET will still be in the same green. But I am worry there might be an edge case that we are not considering now. I will leave the decision to you on whether keeping it or removing it from the check.

Reply:
For the sake of completeness, If EET is greater than cached, that will have already be saved into nearest_green_entry_time so that will be used as the new ET.